### PR TITLE
Faster Typechecking - It's Finished Edition

### PR DESCRIPTION
**Pull request by ike709**
Closes #587

I wanted this to get in before I nuke DreamPath. Credit to @ZeWaka for doing basically all the work.

### CODE REVIEW

File: OpenDreamShared/Dream/DreamPath.cs [hunk 1]
```diff
@@ -135,10 +127,14 @@ public enum PathType {
...
-            // Unsigned overflow is desired, see: Doom3/unity's impl
-            return (typeIndex - ancestor.typeIndex) <= ancestor.numChildren;
+        public bool IsDescendantOf(DreamPath path) {
+            if (path.Elements.Length > Elements.Length) return false;
+
+            for (int i = 0; i < path.Elements.Length; i++) {
+                if (Elements[i] != path.Elements[i]) return false;
+            }
+
+            return true;
```
Comment (ZeWaka) [inline OpenDreamShared/Dream/DreamPath.cs:137]: what's with this?

File: OpenDreamRuntime/Objects/DreamObjectTree.cs [hunk 2]
```diff
@@ -70,6 +78,32 @@ public class TreeEntry {
...
+        /// <summary>
+        /// Returns a list of the TreeEntry's Types sorted in DFS traversal order
+        /// </summary>
+        /// <returns>IEnumerable of sorted TreeEntries</returns>
+        /// <seealso cref="DreamObjectDefinition.IsSubtypeOf">IsDescendantOf</seealso>
+        public IEnumerable<TreeEntry> GetDFSSortedTypeList() {
+            List<TreeEntry> dfs_sorted_types = new();
+            Stack<TreeEntry> dfs_sort_stack = new();
+
+            dfs_sort_stack.Push(Types[0]); // Always the root ('/')
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObjectTree.cs:90]: Add a `public TreeEntry RootType { get => Types[0]; }` property and use that here instead.

File: OpenDreamRuntime/Objects/DreamObjectTree.cs [hunk 3]
```diff
@@ -70,6 +78,32 @@ public class TreeEntry {
...
+            while (dfs_sort_stack.Count != 0) {
+                TreeEntry current = dfs_sort_stack.Pop();
+                dfs_sorted_types.Add(current);
+
+                // Because stacks pop the elements out LIFO, push in reverse
+                // to traverse the returned list in the same order as was returned to us
+                List<int> rev_children = new(current.InheritingTypes);
+                rev_children.Reverse();
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObjectTree.cs:98]: `rev_children` doesn't get used?

File: OpenDreamRuntime/Objects/DreamObjectDefinition.cs [hunk 4]
```diff
@@ -15,6 +15,15 @@ public class DreamObjectDefinition {
         public readonly Dictionary<string, DreamValue> Variables = new();
         public readonly Dictionary<string, int> GlobalVariables = new();
 
+        /// <summary>
+        /// Internal type ID used for performant subclass checking
+        /// </summary>
+        public uint typeIndex = 0;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObjectDefinition.cs:21]: Make this `TypeIndex` to match the names of other public variables.

File: OpenDreamRuntime/Objects/DreamObjectDefinition.cs [hunk 5]
```diff
@@ -15,6 +15,15 @@ public class DreamObjectDefinition {
...
+        /// <summary>
+        /// Number of children for performant subclass checking
+        /// </summary>
+        public uint numChildren = 0;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObjectDefinition.cs:25]: Make this `NumChildren` to match the names of other public variables

File: OpenDreamRuntime/Objects/DreamObject.cs [hunk 6]
```diff
@@ -6,13 +6,15 @@
...
     public class DreamObject {
+        [Dependency] private readonly IDreamManager _dreamMan = default!;
         public DreamObjectDefinition? ObjectDefinition { get; protected set; }
         public bool Deleted = false;
 
         private Dictionary<string, DreamValue> _variables = new();
 
         public DreamObject(DreamObjectDefinition? objectDefinition) {
             ObjectDefinition = objectDefinition;
+            IoCManager.InjectDependencies(this);
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObject.cs:17]: How performant is this?

File: OpenDreamRuntime/Objects/DreamObjectTree.cs [hunk 7]
```diff
@@ -176,7 +210,19 @@ public class TreeEntry {
                 }
             }
 
-            //Third pass: Load each type's vars and procs
+            //Third pass: Create a DFS list of types so that all children of a parent type have contiguous indices,
+            //so we can type-check by just checking in a range. See: DreamPath.IsDescendantOf
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObjectTree.cs:214]:
```suggestion
            //so we can type-check by just checking in a range. See: DreamObjectDefinition.IsSubtypeOf
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Objects/DreamObject.cs (hunk 6 above)
Comment (ike709) [reply OpenDreamRuntime/Objects/DreamObject.cs:17]: DreamObject's constructor isn't even showing up on the profiler. Plus it should cache the injector after the first time.

**Comment by ike709** [comment]
This is horribly broken and I'm not really sure why.

**Comment by ike709** [comment]
Going to bully zewaka to deal with it

**Pull request closed without merging**